From 75c773d73d0f6f893af611dc557a34dff8c7f9d0 Mon Sep 17 00:00:00 2001
From: Chen Jiahao <chenjiahao16@huawei.com>
Date: Fri, 13 May 2022 10:02:18 +0800
Subject: [PATCH 2678/4585] [Huawei] arm: fix vector addr overflow by using
 local symbol names

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 186744

--------------------------------

Spectre-BHB workaround in arm32 introduces exception vectors with
base address 0xffff0000:

ffff1240 t vector_bhb_loop8_irq
ffff1264 t vector_bhb_bpiall_irq
ffff1280 t vector_bhb_loop8_dabt
ffff12a4 t vector_bhb_bpiall_dabt
ffff12c0 t vector_bhb_loop8_pabt
ffff12e4 t vector_bhb_bpiall_pabt
ffff1300 t vector_bhb_loop8_und
ffff1320 t vector_bhb_bpiall_und
ffff1340 t vector_bhb_loop8_fiq
ffff1364 t vector_bhb_bpiall_fiq

which may lead to address overflow when adding a random KASLR offset:

07df1240 t vector_bhb_loop8_irq
07df1278 t vector_bhb_bpiall_irq
07df12a0 t vector_bhb_loop8_dabt
07df12d8 t vector_bhb_bpiall_dabt
07df1300 t vector_bhb_loop8_pabt
07df1338 t vector_bhb_bpiall_pabt
07df1360 t vector_bhb_loop8_und
07df1394 t vector_bhb_bpiall_und
07df13c0 t vector_bhb_loop8_fiq
07df13f8 t vector_bhb_bpiall_fiq

Using .L prefixes to mark them as local symbols, will prevent them
from being visible at all outside the assembler source. This fixes
the problem above.

Fixes: 450c16051983 ("[Backport] ARM: Spectre-BHB workaround")
Signed-off-by: Chen Jiahao <chenjiahao16@huawei.com>
---
 arch/arm/kernel/entry-armv.S | 28 ++++++++++++++--------------
 1 file changed, 14 insertions(+), 14 deletions(-)

diff --git a/arch/arm/kernel/entry-armv.S b/arch/arm/kernel/entry-armv.S
index d74678d9598b..9942ed3ecb37 100644
--- a/arch/arm/kernel/entry-armv.S
+++ b/arch/arm/kernel/entry-armv.S
@@ -1033,7 +1033,7 @@ ENDPROC(.Lvector_\name)
 #ifdef CONFIG_HARDEN_BRANCH_HISTORY
 	.subsection 1
 	.align 5
-vector_bhb_loop8_\name:
+.Lvector_bhb_loop8_\name:
 	.if \correction
 	sub	lr, lr, #\correction
 	.endif
@@ -1049,9 +1049,9 @@ vector_bhb_loop8_\name:
 	dsb
 	isb
 	b	2b
-ENDPROC(vector_bhb_loop8_\name)
+ENDPROC(.Lvector_bhb_loop8_\name)
 
-vector_bhb_bpiall_\name:
+.Lvector_bhb_bpiall_\name:
 	.if \correction
 	sub	lr, lr, #\correction
 	.endif
@@ -1064,7 +1064,7 @@ vector_bhb_bpiall_\name:
 	@ isb not needed due to "movs pc, lr" in the vector stub
 	@ which gives a "context synchronisation".
 	b	2b
-ENDPROC(vector_bhb_bpiall_\name)
+ENDPROC(.Lvector_bhb_bpiall_\name)
 	.previous
 #endif
 
@@ -1236,24 +1236,24 @@ ENDPROC(vector_bhb_bpiall_\name)
 	.section .vectors.bhb.loop8, "ax", %progbits
 .L__vectors_bhb_loop8_start:
 	W(b)	.Lvector_rst
-	W(b)	vector_bhb_loop8_und
+	W(b)	.Lvector_bhb_loop8_und
 	W(ldr)	pc, .L__vectors_bhb_loop8_start + 0x1004
-	W(b)	vector_bhb_loop8_pabt
-	W(b)	vector_bhb_loop8_dabt
+	W(b)	.Lvector_bhb_loop8_pabt
+	W(b)	.Lvector_bhb_loop8_dabt
 	W(b)	.Lvector_addrexcptn
-	W(b)	vector_bhb_loop8_irq
-	W(b)	vector_bhb_loop8_fiq
+	W(b)	.Lvector_bhb_loop8_irq
+	W(b)	.Lvector_bhb_loop8_fiq
 
 	.section .vectors.bhb.bpiall, "ax", %progbits
 .L__vectors_bhb_bpiall_start:
 	W(b)	.Lvector_rst
-	W(b)	vector_bhb_bpiall_und
+	W(b)	.Lvector_bhb_bpiall_und
 	W(ldr)	pc, .L__vectors_bhb_bpiall_start + 0x1008
-	W(b)	vector_bhb_bpiall_pabt
-	W(b)	vector_bhb_bpiall_dabt
+	W(b)	.Lvector_bhb_bpiall_pabt
+	W(b)	.Lvector_bhb_bpiall_dabt
 	W(b)	.Lvector_addrexcptn
-	W(b)	vector_bhb_bpiall_irq
-	W(b)	vector_bhb_bpiall_fiq
+	W(b)	.Lvector_bhb_bpiall_irq
+	W(b)	.Lvector_bhb_bpiall_fiq
 #endif
 
 	.data
-- 
2.17.1

